theopensystemslab/planx-new
# feat: create a ContactInput component type

**Changes:**
Introduces a `ContactInput` component type which will replace individual text inputs used to collect appplicant names, phone numbers, emails, etc.

The passport representation for applicants (or agents, proxies, etc) changes as follows when switching from many `TextInput` components to a single `ContactInput` component and has been updated in the BOPS & Uniform schema mappings (the schemas will support _both_ so there's no strict adoption rule/timeline of the new component):
Before:
```json
"passport": {
  "data": {
    "applicant.agent.name.first",
    "applicant.agent.name.last",
    "..."
   }
 }
 ```
 
 After: 
 ```json
"passport": {
   "data": {
     "applicant.agent": {
       "name.first",
       "name.last",
       "..."
    }
   }
 }
 ```
**How to test:**
- I've updated this pizza's `about-the-applicant` flow to use the new component: https://1207.planx.pizza/opensystemslab/about-the-applicant
![Screenshot from 2022-11-04 12-28-32](https://user-images.githubusercontent.com/5132349/199962463-f71c0c81-c241-43c0-bd54-4f11395a76a0.png)
- I've published a parent LDC flow to capture that change: https://1207.planx.pizza/buckinghamshire/apply-for-a-lawful-development-certificate/preview?analytics=false
- You should go through the LDC flow, confirm that you see the new ContactInput components and that your passport is consistent with how it used to be, and that all necessary applicant fields are still sent to BOPS & Uniform as expected

---
**Prior context:**
One of a couple experiments this week related to navigation (see also #1209). This one aims to address a very common use case which is entering your contact details, currently spread out over ~6 individual text inputs: https://editor.planx.uk/opensystemslab/about-the-applicant. Largely follows the same structure as our existing `AddressInput` component.

Pros:
- Maintains accessibilty
- Doesn't touch graph logic
- Quick to rollout

Cons: 
- Lack of flexibility / editor control
- Loss of individual "description" to supplement input label
- Is a bandaid to the dream feature

Todos and cleanup if moving forward (estimate 1):
- [x] Add valid email regex (see `TextInput` model)
- [x] Clean up `SummaryList`/`Review` page presentation
- [x] Ensure passport variables will match existing flow, check if any data models need to be adjusted for CSV/BOPS/Uniform
